plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataBuilder.java

```diff
@@ -0,0 +1,356 @@
+    return true;
+  }
+
+  // -----------------------------------
+  // artifactId
+
+  public boolean onEnterArtifactId(final String artifactId) {
+    if (setArtifactId(artifactId)) {
+      log.debug("-> GA: {}:{}", getGroupId(), getArtifactId());
+      baseVersions.clear();
+      return true;
+    }
+    return false;
+  }
+
+  @Nullable
+  public Maven2Metadata onExitArtifactId() {
+    checkState(getArtifactId() != null, "artifactId");
+    log.debug("<- GA: {}:{}", getGroupId(), getArtifactId());
+    if (baseVersions.isEmpty()) {
+      log.debug("Nothing to generate: {}:{}", getGroupId(), getArtifactId());
+      return null;
+    }
+    Iterator<Version> vi = baseVersions.descendingIterator();
+    String latest = vi.next().toString();
+    String release = latest;
+    while (release.contains("SNAPSHOT") && vi.hasNext()) {
+      release = vi.next().toString();
+    }
+    if (release.contains("SNAPSHOT")) {
+      release = null;
+    }
+    return Maven2Metadata.newArtifactLevel(
+        DateTime.now(),
+        getGroupId(),
+        getArtifactId(),
+        latest,
+        release,
+        Iterables.transform(baseVersions, new Function<Version, String>()
+        {
+          @Override
+          public String apply(final Version input) {
+            return input.toString();
+          }
+        }));
+  }
+
+  private void addBaseVersion(final String baseVersion) {
+    checkNotNull(baseVersion, "baseVersion");
+    try {
+      if (baseVersions.add(versionScheme.parseVersion(baseVersion))) {
+        log.debug("Added base version {}:{}:{}", getGroupId(), getArtifactId(), baseVersion);
+      }
+    }
+    catch (InvalidVersionSpecificationException e) {
+      // According to versionScheme implementation we use, this exception will never happen
+      // log + ignore
+      log.info("Invalid baseVersion discovered: " + baseVersion , e);
+    }
+  }
+
+  // -----------------------------------
+  // baseVersion
+
+  /**
+   * Internal structure to hold parsed Aether {@link Version} and {@link Coordinates}.
+   */
+  private static class VersionCoordinates
+  {
+    private final Version version;
+
+    private final Coordinates coordinates;
+
+    private VersionCoordinates(final Version version, final Coordinates coordinates) {
+      this.version = version;
+      this.coordinates = coordinates;
+    }
+  }
+
+  public boolean onEnterBaseVersion(final String baseVersion) {
+    if (setBaseVersion(baseVersion)) {
+      log.debug("-> GAbV: {}:{}:{}", getGroupId(), getArtifactId(), getBaseVersion());
+      latestVersionCoordinatesMap.clear();
+      latestVersionCoordinates = null;
+      return true;
+    }
+    return false;
+  }
+
+  @Nullable
+  public Maven2Metadata onExitBaseVersion() {
+    checkState(getBaseVersion() != null, "baseVersion");
+    log.debug("<- GAbV: {}:{}:{}", getGroupId(), getArtifactId(), getBaseVersion());
+    if (!getBaseVersion().endsWith("SNAPSHOT") || latestVersionCoordinates == null) {
```

Comment:
elsewhere we're using `contains("SNAPSHOT")` to check for snapshots - an `isSnapshot` utility method would be useful